### PR TITLE
feat: Introduce the `FixedResolutionViewport`

### LINKED ISSUE #2776: FixedResolutionViewport on the new CameraComponent API
# Problem to solve
In the old camera API, we can use `FixedResolutionViewport` to set a virtual size. 
```
camera.viewport = FixedResolutionViewport(Vector2(1280, 720));
```

In the new camera API we can use `CameraComponent.withFixedResolution`  that internally looks like:
```
CameraComponent(
      world: world,
      viewport: FixedAspectRatioViewport(aspectRatio: width / height)
        ..addAll(hudComponents ?? []),
      viewfinder: Viewfinder()..visibleGameSize = Vector2(width, height),
    )
```

Using `CameraComponent.withFixedResolution` will not make the viewport to a fixed resolution. It partially works because the `camera.viewfinder.visibleWorldRect` will always be fixed to the given `width` and `height` but `camera.viewport.size` is not fixed and, also, `zoom` is not working at all. 

-----
When using `Forge2D` in the old API I can have a  viewport with `FixedResolutionViewport(Vector2(1280, 720))`. I can add all the UI components to this viewport using `PositionType.viewport`. 

For the Forge2D components, I will set a `zoom` of `100` to scale the world to a resolution of `12.8 * 7.2` to overcome Box2D speed and size limitations, and I can add all these components using the scaled values. 

Here is a git repository with an example  of how I use the old `FixedResolutionViewPort`
https://github.com/Yayo-Arellano/flutter_games_compilation/blob/main/flutter_learn_flame/lib/my_game.dart#L31

### PR DESCRIPTION
**Pull request by spydon**
# Description
This enables you to zoom when using the `CameraComponent.withFixedResolution` constructor, without disabling the fixed resolution after doing so, by introducing a new viewport the `FixedResolutionViewport`.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] No, this PR is not a breaking change.

Maybe slightly breaking for the people that have done work-arounds.

## Related Issues

Closes #2776

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by ufrshubham** [comment]
Reposting here:
Components attached to viewfinder look correct now. They work like HUD and maintain their scale and position w.r.t to the virual resolution as expected. But the gesture events seem wrong. In the following video you can see that the pause button detects tap events assuming the button to be in the world space.

https://github.com/flame-engine/flame/assets/33748002/2370582d-8a39-4ab7-865f-a7ff8c652fe1

**Comment by spydon** [comment]
This has shown quite tricky to get clean code for, so I'm turning it to a draft.
I'm currently experimenting with introducing a Lens beside the Viewfinder and Viewport.

File: packages/flame/lib/src/camera/viewports/fixed_aspect_ratio_viewport.dart [hunk 1]
```diff
@@ -37,6 +37,7 @@ class FixedAspectRatioViewport extends Viewport {
     _handleResize(size);
   }
 
+  @protected
```
Comment (erickzanardo) [inline packages/flame/lib/src/camera/viewports/fixed_aspect_ratio_viewport.dart:40]: Wait, why is this needed? This is already private

Comment (spydon) [reply packages/flame/lib/src/camera/viewports/fixed_aspect_ratio_viewport.dart:40]: Haha, just something that was left in-between all the refactors, removing it.

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/camera/camera_component.dart [hunk 2]
```diff
@@ -70,15 +68,16 @@ class CameraComponent extends Component {
...
     required double height,
+    Viewfinder? viewfinder,
     World? world,
     Component? backdrop,
     List<Component>? hudComponents,
   }) {
     return CameraComponent(
       world: world,
-      viewport: FixedAspectRatioViewport(aspectRatio: width / height)
+      viewport: FixedResolutionViewport(resolution: Vector2(width, height))
```
Comment (luanpotter) [inline packages/flame/lib/src/camera/camera_component.dart:78]: i'm so excited to have this!

File: packages/flame/lib/src/camera/viewfinder.dart [hunk 3]
```diff
@@ -133,25 +137,30 @@ class Viewfinder extends Component
...
+      _updateZoom();
     }
   }
 
   /// See [CameraComponent.visibleWorldRect].
   @internal
-  Rect get visibleWorldRect => _visibleRect ??= _computeVisibleRect();
-  Rect? _visibleRect;
-  Rect _computeVisibleRect() {
+  Rect get visibleWorldRect => visibleRect ??= computeVisibleRect();
```
Comment (luanpotter) [inline packages/flame/lib/src/camera/viewfinder.dart:146]: could this be a useful getter to expose to users?
sounds like something people might want to access for certain things

File: examples/lib/stories/camera_and_viewport/fixed_resolution_example.dart [hunk 4]
```diff
@@ -12,32 +13,84 @@ class FixedResolutionExample extends FlameGame
...
+        text: 'Viewfinder\ncomponent',
+        textRenderer: textRenderer,
+        position: Vector2(0, 200),
+        anchor: Anchor.center,
+      ),
+    );
+    camera.viewport.add(
+      TextButton(
+        text: 'Viewport\ncomponent',
+        position: camera.viewport.virtualSize - Vector2.all(10),
```
Comment (luanpotter) [inline examples/lib/stories/camera_and_viewport/fixed_resolution_example.dart:50]: is this basically always 600x1024 (for example?)

File: packages/flame/lib/src/camera/viewport.dart [hunk 5]
```diff
@@ -64,6 +66,8 @@ abstract class Viewport extends Component
     return _size;
   }
 
+  Vector2 get virtualSize => size;
```
Comment (luanpotter) [inline packages/flame/lib/src/camera/viewport.dart:69]: quick docs?

File: examples/lib/stories/camera_and_viewport/fixed_resolution_example.dart (hunk 4 above)
Comment (spydon) [reply examples/lib/stories/camera_and_viewport/fixed_resolution_example.dart:50]: Yeah, now it is, it was configurable before but since it is now using `.controlled` it doesn't update.

**Review by luanpotter — APPROVED** [review]
😍

File: packages/flame/lib/src/camera/viewfinder.dart (hunk 3 above)
Comment (spydon) [reply packages/flame/lib/src/camera/viewfinder.dart:146]: It is exposed through the `CameraComponent`

File: packages/flame/lib/src/camera/viewport.dart (hunk 5 above)
Comment (spydon) [reply packages/flame/lib/src/camera/viewport.dart:69]: Oh yes!
